homeassistant/components/texecom.py

```diff
@@ -0,0 +1,222 @@
+        """Initialize the device."""
+        self._number = zone_number
+        self._name = zone_name
+        self._sensor_type = zone_type
+        self._state = state
+
+        _LOGGER.debug('Setting up zone: %s', self._number)
+        _LOGGER.debug('Setting up zone: %s', self._name)
+        _LOGGER.debug('Setting up zone: %s', self._sensor_type)
+        _LOGGER.debug('Setting up zone: %s', self._state)
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        async_dispatcher_connect(
+            self.hass, SIGNAL_ZONE_UPDATE, self._update_callback)
+
+    @property
+    def device_class(self):
+        """Return the class of this sensor."""
+        return self._sensor_type
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def is_on(self):
+        """Return the name of the device."""
+        return self._state
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @callback
+    def _update_callback(self, data):
+        """Update the zone's state, if needed."""
+        _LOGGER.debug('Attempting to Update Zone %s', self._name)
+
+        if self._number == data.signalledzone:
+            _LOGGER.info('Correct zone found to update %s', self._name)
+            _LOGGER.debug('The new state is %s', data.zonestate)
+
+            if data.zonestate == '0':
+                _LOGGER.debug('Setting zone state to false')
+                self._state = False
+            elif data.zonestate == '1':
+                _LOGGER.debug('Setting zone state to true')
+                self._state = True
+            else:
+                _LOGGER.debug('Unknown state assuming tamper')
+                self._state = True
+
+            _LOGGER.info('New Zone State is %s', self._state)
+            self.async_schedule_update_ha_state()
+
+
+class TexecomPanelInterface(Entity):
+    """Representation of a Texecom Panel Interface."""
+
+    def __init__(self, name, port):
+        """Initialize the Texecom Panel Interface."""
+        self._name = name
+        self._state = None
+        self._port = port
+        self._baudrate = '19200'
+        self._serial_loop_task = None
+        self._attributes = []
+        self.signalledzone = '0'
+        self.zonestate = '0'
+
+        _LOGGER.info('Setting up Serial: %s', name)
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Handle when an entity is about to be added to Home Assistant."""
+        self._serial_loop_task = self.hass.loop.create_task(
+            self.serial_read(self._port, self._baudrate))
+
+    @asyncio.coroutine
+    def serial_read(self, device, rate, **kwargs):
+        """Read the data from the port."""
+        import serial_asyncio
+        _LOGGER.info('Opening Serial Port')
+        reader, _ = yield from serial_asyncio.open_serial_connection(
```

Comment:
Device specific I/O and data parsing should be moved to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services